osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
+      owner_tid_ = my_tid_;
+      owner_tid_is_valid_ = true;
+      locked_ = true;
+    }
+    return locked_;
+  }
+
+  ~RocksDBGuard() {
+    if (locked_) {
+      owner_tid_is_valid_ = false;
+      mu_.unlock();
+    }
+  }
+
+ private:
+  int my_tid_;
+  bool locked_;
+
+  static std::mutex mu_;
+  static int owner_tid_;
+  // linux has no notion of an invalid thread id, hence owner_tid_is_valid_
+  static bool owner_tid_is_valid_;
```

Comment:
No. The value of owner_tid_ is a valid thread id if and only if owner_tid_is_valid_ is true. Otherwise, owner_tid_ does not contain a valid thread it. This mechanism is necessary because pthread does not provide a reserved value to set as a distinct invalid thread id. 

owner_thread_is_valid_ will be false when no one owns the thread. My next comment has a reference to a discussion on this...